Adds transaction example

s-ben opened the pull request:
Adds an example transaction to the [Transaction Details](https://docs.decred.org/advanced/transaction-details/) page ( Closes #765 ). 

Note: the transaction fields were broken down according to the example tx provided in #765 by @davecgh, and when parsing the data I ended up moving a couple bytes around. E.g. `script version` in the example was `0000`, but the docs say this is a two byte field, so have used `00` and moved the subsequent `00` into the next field (`script length`). There are also a couple fields that need double checking, such as `public key signature length` (hex evaluating to 25, though looks like the script has 50 bytes?) and `signature script length`. 

Also, the markdown table is a bit wonky, due to the fact that markdown doesn't support spanning cells. Can tweak the formatting with inline html, but wanted to get a first pass review on the current table first. We could also just use text formatting like we do for the block header [example](https://docs.decred.org/advanced/block-header-specifications/).

Also, do we want to paste a link to the full decoded tx JSON?
https://explorer.dcrdata.org/api/tx/decoded/a58389bd42520a9b9ecbae5432e27b91fde74d2f8e74ac6f8c22cda57b8ae348?indent=true

davecgh commented:
Each hex digit is a nibble (4 bits), so 2 bytes takes 4 hex digits.  The version is indeed `0000`.

s-ben commented:
So how is the verision = `1` in the decoded [JSON](https://explorer.dcrdata.org/api/tx/decoded/a58389bd42520a9b9ecbae5432e27b91fde74d2f8e74ac6f8c22cda57b8ae348?indent=true) from the explorer,
```
{
   "txid": "a58389bd42520a9b9ecbae5432e27b91fde74d2f8e74ac6f8c22cda57b8ae348",
   "version": 1,
```
I believe this was my reasoning/rationalization for my guesswork.

davecgh commented:
That's the transaction version.  The script version is a different version.

davecgh commented:
A couple of things:

* In the transaction format section, each script should be preceded with a script length field which is itself a variable number of bytes
* The values are all raw hex values and, as such, imo, shouldn't have additional commas which I find confusing

s-ben commented:
> In the transaction format section, each script should be preceded with a script length field which is itself a variable number of bytes

Added variable to table:
`Signature script length | The length of the signature script as a variable-length integer` 

> The values are all raw hex values and, as such, imo, shouldn't have additional commas which I find confusing

I found that confusing as well, but was just following the convention in the [block header example](https://docs.decred.org/advanced/block-header-specifications/). Do we want to get rid of the commas there too?

s-ben commented:
@davecgh with the changes made per above comment, I think we're close to a workable example. However, wondering if more can/should be documented here. Personally, I'd like to understand Decred's transactions at a lower level. E.g. how the variable length integers are parsed. I think this could be valuable to other developers as well. But I'm wondering what questions we're getting from developers specifically, and what are they trying to do generally. Presumably they need to be able to a) parse transactions and b) construct transactions. But what types of transactions? At what level of abstraction? Do we need to document stakebase and other types of transactions? What is Decred-specific, and what is already covered by the Bitcoin dev docs [transaction guide](https://bitcoin.org/en/transactions-guide#introduction)? Also, do we want to do a diagram, like this one from the Bitcoin dev guide (which is an open source SVG if we want to start with this)?

![image](https://user-images.githubusercontent.com/1634777/57565465-3b030e00-7373-11e9-8f63-28fdd3383715.png)

davecgh commented:
> I found that confusing as well, have removed. Was just following the convention in the [block header example](https://docs.decred.org/advanced/block-header-specifications/). Do we want to get rid of the commas there too?

Commas are acceptable between each byte (2 hex digits), but there was something strange about how they were formatted such that is was non-standard and confusing.

davecgh commented:
An SVG diagram would be helpful I think.  As far as the other questions, all transactions follow the format specified here.  Votes, Ticket Purchases, etc, are merely specific incantations of them.

s-ben commented:
I believe this (public domain) diagram from the Bitcoin wiki could be easily modified for our purposes,
https://en.bitcoin.it/wiki/File:TxBinaryMap.png

I am wondering if this text from the diagram is true for Decred as well?
> Scripts and DER encoding both use big-endian values, all other serializations use little-endian

Also, this may be totally enough for this page. However, when I pretend to be a developer looking at this, I still have questions. What are the rules for encoding different types of transactions? What is the byte map for specific types of scripts? Are they the same as Bitcoin, or are they different? Can I just use Bitcoin documentation for this? What types of scripts are supported? What is the byte map for Decred-specific transactions (e.g. transactions for buying, voting, and revoking tickets)? Do they follow some rules? Do I just need to look to the source code for answers to these questions?

davecgh commented:
The bit about script and DER encoding using big-endian values is correct.  Regarding the rest, that is way too much to put into a general transaction specification imo.  What you're discussing is separate sections which discuss the various types of *standard* transactions.

s-ben commented:
> The bit about script and DER encoding using big-endian values is correct.

Tried and failed to come up with a simple way to word this. Realized that this is already painstakingly documented for Bitcoin transactions, and that if we're not doing anything different (at this level of abstraction anyway), it's probably not worth re-documenting here.

> Regarding the rest, that is way too much to put into a general transaction specification imo. What you're discussing is separate sections which discuss the various types of standard transactions and scripts.

For this page, you're probably right. I do still think it's useful to provide more documentation/explanation of Decred-specific transactions (e.g. stakebase transactions). But that doesn't need to be this page.

Perhaps for this PR, we should just stop here and do a final round of review on what we have so far?
* Updating `Transaction Format` section with `Signature script length` variable
* Adding transaction example

davecgh commented:
I agree that it would be useful to document each one for sure, but I just don't think it belongs in the general transaction specification.  They are strictly subsets of generalized transactions.

davecgh commented:
In terms of review, the section at the bottom doesn't appear formatted correctly.

s-ben commented:
Agreed, we probably don't want to put any more detail in the general transaction specification. Have proposed documenting Decred-specific transactions further in an [Issue](https://github.com/decred/dcrdocs/issues/942) to gather feedback on the idea. 

Table formatting is fixed. Markdown formatting is limited (e.g. no merged cells) so had to convert to html to get something more readable.

davecgh commented:
Already discussed on matrix, but the table formatting is still not correct on some browsers such as Firefox.

s-ben commented:
Ok fixed table formatting for Firefox. Turns out this is a common problem, which gets into some thorny CSS issues. I ended up just hard-coding the text wrapping for cells with long strings; the other, more elegant solutions from SO weren't working. Is this hacky CSS OK? Have tested on Brave, Chrome, Firefox and Safari and rendering fine on all now.

davecgh commented:
Still looks broken to me in FF 66.0.5.

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Expiry</td>
+<td><code>00000000</code></td>
+<td>0</td>
```
The description is correct for the number, but it might be worth adding that 0 means there is no expiry.

s-ben replied:
Noted:
```
0
(no expiry)
```

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Lock time</td>
+<td><code>00000000</code></td>
+<td>0</td>
```
The description is correct for the number, but it might be worth adding that 0 means there is no lock time.

s-ben replied:
Noted:
```
0
(no lock time)
```

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<td colspan="3" style="text-align: center;">1st witness</td>
+</tr>
+<tr>
+  <td>Input amount</td>
```
Extra spaces here that are nowhere else.

s-ben replied:
Removed

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<td colspan="3" style="text-align: center;">2nd input</td>
+</tr>
+<tr>
+  <td>Transaction hash</td>
```
Extra spaces not anywhere else.

s-ben replied:
Removed

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Output amount</td>
+<td><code>91c2e26d02000000</code></td>
+<td>104.33</td>
```
104.33512081

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Output amount</td>
+<td><code>d676dd6801000000</code></td>
+<td>60.54</td>
```
60.54311638

s-ben replied:
Updated all numbers to 8 decimal places.

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+  <td>Input amount</td>
+<td><code>ca68865402000000</code></td>
+<td>100.08</td>
```
100.08029386

s-ben replied:
Updated

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Input amount</td>
+<td><code>49743a8201000000</code></td>
+<td>64.79</td>
```
64.79836233

s-ben replied:
Updated

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Signature script</td>
+<td style="word-break: break-all;"><code>4730440220139466bd10b1f071f9b4ac16ed434d63d090613fd06470f7bca9bba2b6ea6a0a02203cad1297ab9384466b3f304cd61e7e4ffa9abc1ae9fe1630207fae6f200da1680121024d540859b805c5780f2f0da350df8757c1defb72e3381b252d20874478a5549c</code></td>
+<td></td>
```
2 data pushes.  1st push is DER signature.  2nd push is public key.

s-ben replied:
Just used this text as is. This is definitely helpful, though wondering, would it be useful to break this down further? E.g. separate out DER signature and public key hex values and label them? Or can we assume the reader knows how to interpret this?

davecgh replied:
If it will fit well, that makes sense.  I was trying to keep it relatively short because the format I see on the current markdown preview is wonky.

If you want to break it down more though, it's:

2 data pushes.

| Script Operation | Value | Description|
| --- | --- | --- |
| 1st push |`3045022100da5b0fde58c4c57a88d96f83b02eddcb67b1b68ce95cc562c895f2e4e` `57bb44302201a8744aed654d2979bdd797187f6bf63ad1898c5d5c585309eeca558` `742858b201` | DER signature |
| 2nd push | `0307e3d98b004b15561c0f5aa158c90f93de33756d3bfd1f8a1058a4d90ffc7ec7` | Public key |

s-ben replied:
Formatting is getting a little wonky, but my guess is that if you're slogging through this level of detail, you'll probably appreciate more breakdown. 

Below is a screenshot of how I've implemented it (FF),
![image](https://user-images.githubusercontent.com/1634777/57825146-7e35f600-7752-11e9-9402-4441d172924b.png)

Let me know if this text is inaccurate, or could be worded more clearly. I'm still slightly confused as to what the DER signature is. I assume it's a binary string of digits encoded with DER, which is what `asm` is in the decoded JSON (i.e. this is still encoded),

```
         "scriptSig": {
            "asm": "3045022100da5b0fde58c4c57a88d96f83b02eddcb67b1b68ce95cc562c895f2e4e57bb44302201a8744aed654d2979bdd797187f6bf63ad1898c5d5c585309eeca558742858b201 0307e3d98b004b15561c0f5aa158c90f93de33756d3bfd1f8a1058a4d90ffc7ec7",
            "hex": "483045022100da5b0fde58c4c57a88d96f83b02eddcb67b1b68ce95cc562c895f2e4e57bb44302201a8744aed654d2979bdd797187f6bf63ad1898c5d5c585309eeca558742858b201210307e3d98b004b15561c0f5aa158c90f93de33756d3bfd1f8a1058a4d90ffc7ec7"
         }
```

davecgh replied:
That text is fine except it's just a DER signature.  No need to say asm.  asm just means it's the disassembly of the script.

There really is no need to break the DER signature down even more in this document imo.  It's a well known and defined ASN.1 encoding.

s-ben replied:
Yes, I think we have drilled far enough. Removed `(asm)`.

davecgh commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Signature script</td>
+<td style="word-break: break-all;"><code>483045022100da5b0fde58c4c57a88d96f83b02eddcb67b1b68ce95cc562c895f2e4e57bb44302201a8744aed654d2979bdd797187f6bf63ad1898c5d5c585309eeca558742858b201210307e3d98b004b15561c0f5aa158c90f93de33756d3bfd1f8a1058a4d90ffc7ec7</code></td>
+<td></td>
```
2 data pushes.  1st push is DER signature.  2nd push is public key.

s-ben replied:
Same comment as above

s-ben replied:
Same comment as above. Note that the data and from above was from the 2nd signature script, even though comment was on the 1st. 

Tried to follow what you did when breaking down the other signature. Should definitely be reviewed for accuracy though.

davecgh replied:
Reviewed.  It's accurate.

s-ben commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Sequence number</td>
+<td><code>ffffffff</code></td>
+<td>4294967295</td>
```
Added `(transaction finalized)`

s-ben commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Tree</td>
+<td><code>00</code></td>
+<td>0</td>
```
Added `(use regular transaction fee)`

davecgh replied:
tree, not fee

s-ben replied:
Fixed

s-ben commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<td>104.33</td>
+</tr>
+<tr>
+<td>Version</td>
```
Added

s-ben commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Public key script</td>
+<td style="word-break: break-all;"><code>76a914f2ccddc70f8b8bcea24a362bf404841d53cbebc088ac</code></td>
+<td></td>
```
Added. Used this format,
`OP_DUP OP_HASH160 f2ccddc70f8b8bcea24a362bf404841d53cbebc0 OP_EQUALVERIFY OP_CHECKSIG` 
`(Pay-To-Pubkey (P2Pk) hash script)`

davecgh replied:
That's fine, but it needs to be `(Pay-To-Pubkey-Hash (P2PKH) script)`.  It's actually paying to the pubkey hash, not the pubkey.

s-ben replied:
Updated

s-ben commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<tr>
+<td>Public key script</td>
+<td style="word-break: break-all;"><code>76a9141a1f5c7e9f7696989dced5cb9b61464f95790b7288ac</code></td>
+<td></td>
```
Added. Same format as above comment.

davecgh replied:
Same here regarding paying to the pubkey hash.

s-ben replied:
Fixed

s-ben commented:
@davecgh have implemented all your suggested changes, as well as tweaked the formatting to accommodate some of the new long strings (working on FF as well). My only outstanding question is on the Signature Script description (in comment on suggestion).

davecgh commented:
I still see several unaddressed review items.  They are collapsed.

s-ben commented:
> I still see several unaddressed review items. They are collapsed.

Are there wrongly resolved comments? Have addressed the outstanding comments I can see.

davecgh commented:
Reviewed updates, resolved relevant comments, and answered questions.

If you look, there is a "7 hidden conversations" in the middle where the review comments are hidden until you click the "Load more...".  Those are the unaddressed items I was referring to.

s-ben commented on docs/advanced/transaction-details.md:
```diff
@@ -75,3 +76,210 @@ The witness data of a transaction involves only its inputs. The included data fi
+<td>Public key script length</td>
+<td><code>19</code></td>
+<td>25</td>
+<td></td>
```
Removed

s-ben commented:
Found the "7 hidden conversations" in the middle, but when I click on them it just expands to show already resolved changes :/ Do you see any unresolved changes still on your end? I hate this UI feature.

davecgh commented:
I resolved them because you addressed them in the latest commit.  I approved the PR already.

s-ben commented:
Cool. Thx for the thorough review/patience on this one 👍
